sidecar/src/one_way_shared_memory.rs

```diff
@@ -0,0 +1,207 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present Datadog, Inc.
+
+use datadog_ipc::platform::{FileBackedHandle, MappedMem, NamedShmHandle, ShmHandle};
+use std::ffi::CString;
+use std::io;
+use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+use std::sync::Mutex;
+
+pub struct OneWayShmWriter<T>
+where
+    T: FileBackedHandle + From<MappedMem<T>>,
+{
+    handle: Mutex<Option<MappedMem<T>>>,
+}
+
+pub struct OneWayShmReader<T, D>
+where
+    T: FileBackedHandle + From<MappedMem<T>>,
+{
+    handle: Option<MappedMem<T>>,
+    current_data: Option<Vec<u64>>,
+    pub extra: D,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+struct RawMetaData {
+    generation: AtomicU64,
+    size: usize,
+    writing: AtomicBool,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+struct RawData {
+    meta: RawMetaData,
+    buf: [u8],
+}
+
+impl RawData {
+    fn as_slice(&self) -> &[u8] {
+        // Safety: size is expected to be truthful
+        unsafe { std::slice::from_raw_parts(self.buf.as_ptr(), self.meta.size) }
+    }
+
+    fn as_slice_mut(&mut self) -> &mut [u8] {
+        // Safety: size is expected to be truthful
+        unsafe { std::slice::from_raw_parts_mut(self.buf.as_mut_ptr(), self.meta.size) }
+    }
+}
+
+impl From<&[u64]> for &RawData {
+    fn from(value: &[u64]) -> Self {
+        // Safety: MappedMem is supposed to be big enough
+        // Safety: u64 is aligned
+        unsafe { &*(value as *const [u64] as *const RawData) }
+    }
+}
+
+// Safety: Caller needs to ensure the u8 is 8 byte aligned
+unsafe fn reinterpret_u8_as_u64_slice(slice: &[u8]) -> &[u64] {
+    // Safety: given 8 byte alignment, it's guaranteed to be readable
+    std::slice::from_raw_parts(slice.as_ptr() as *const u64, (slice.len() + 7) / 8)
+}
+
+pub fn create_anon_pair() -> anyhow::Result<(OneWayShmWriter<ShmHandle>, ShmHandle)> {
+    let handle = ShmHandle::new(0x1000)?;
+    Ok((
+        OneWayShmWriter {
+            handle: Mutex::new(Some(handle.clone().map()?)),
+        },
+        handle,
+    ))
+}
+
+impl<T: FileBackedHandle + From<MappedMem<T>>, D> OneWayShmReader<T, D> {
+    pub fn new(handle: Option<MappedMem<T>>, extra: D) -> OneWayShmReader<T, D> {
+        OneWayShmReader {
+            handle,
+            current_data: None,
+            extra,
```

Comment:
Is extra going to be anything else than an endpoint?

Reply:
No, effectively not (as it stands now), but I need two different "types" of endpoints at least to derive the correct path from the endpoint, for different usages of it (e.g. agent config must not write to the same file than actual remote config).